File: 2/Permutations.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

class Permutations {

	public static List<List<Integer>> getPermutations(Integer arr[]) {
		List<List<Integer>> res = new ArrayList<>();
		getPermutationsHelper(0, arr, res);
		return res;
	}

	public static void getPermutationsHelper(int i, Integer arr[], List<List<Integer>> res) {
		if(i == arr.length - 1) {
			res.add(List.of(arr.clone()));
			return;
		}

		for(int j=i; j<arr.length; j++) {
			swap(i, j, arr);
			getPermutationsHelper(i+1, arr, res);
			swap(i, j, arr);
		}
	}

	public static void swap(int i, int j, Integer arr[]) {
		int temp = arr[i];
		arr[i] = arr[j];
		arr[j] = temp;
	}

	public static void main(String args[]) {
		Integer arr[] = {1,2,3};
		System.out.println(getPermutations(arr));
	}
}

File: 2/js/src/Permutations.ts
function swap(i: number, j: number, list: number[]) {
	[list[i], list[j]] = [list[j], list[i]];
}

function getPermutations(list: number[]): number[][] {
	const permutations: number[][] = [];
	getPermutationsHelper(0, list, permutations);
	return permutations;
}

function getPermutationsHelper(i: number, list: number[], permutations: number[][]) {
	if(i == list.length - 1) {
		permutations.push(Array.from(list));
	}

	for(let j=i; j<list.length; j++) {
		swap(i, j, list);
		getPermutationsHelper(i+1, list, permutations);
		swap(i, j, list);
	}
}

(function main() {
	const list = [1,2,3];

	console.log(getPermutations(list));
})();
